src/content/accept_encoding.rs

```diff
@@ -0,0 +1,429 @@
+/// use http_types::Response;
+///
+/// let mut accept = AcceptEncoding::new();
+/// accept.push(EncodingProposal::new(Encoding::Brotli, Some(0.8))?);
+/// accept.push(EncodingProposal::new(Encoding::Gzip, Some(0.4))?);
+/// accept.push(EncodingProposal::new(Encoding::Identity, None)?);
+///
+/// let mut res = Response::new(200);
+/// let encoding = accept.negotiate(&[Encoding::Gzip, Encoding::Brotli])?;
+/// encoding.apply(&mut res);
+///
+/// assert_eq!(res["Content-Encoding"], "br");
+/// #
+/// # Ok(()) }
+/// ```
+pub struct AcceptEncoding {
+    wildcard: bool,
+    entries: Vec<EncodingProposal>,
+}
+
+impl AcceptEncoding {
+    /// Create a new instance of `AcceptEncoding`.
+    pub fn new() -> Self {
+        Self {
+            entries: vec![],
+            wildcard: false,
+        }
+    }
+
+    /// Create an instance of `AcceptEncoding` from a `Headers` instance.
+    pub fn from_headers(headers: impl AsRef<Headers>) -> crate::Result<Option<Self>> {
+        let mut entries = vec![];
+        let headers = match headers.as_ref().get(ACCEPT_ENCODING) {
+            Some(headers) => headers,
+            None => return Ok(None),
+        };
+
+        let mut wildcard = false;
+
+        for value in headers {
+            for part in value.as_str().trim().split(',') {
+                let part = part.trim();
+
+                // Handle empty strings, and wildcard directives.
+                if part.is_empty() {
+                    continue;
+                } else if part == "*" {
+                    wildcard = true;
+                    continue;
+                }
+
+                // Try and parse a directive from a str. If the directive is
+                // unkown we skip it.
+                if let Some(entry) = EncodingProposal::from_str(part)? {
+                    entries.push(entry);
+                }
+            }
+        }
+
+        Ok(Some(Self { entries, wildcard }))
+    }
+
+    /// Push a directive into the list of entries.
+    pub fn push(&mut self, prop: impl Into<EncodingProposal>) {
+        self.entries.push(prop.into());
+    }
+
+    /// Returns `true` if a wildcard directive was passed.
+    pub fn wildcard(&self) -> bool {
+        self.wildcard
+    }
+
+    /// Set the wildcard directive.
+    pub fn set_wildcard(&mut self, wildcard: bool) {
+        self.wildcard = wildcard
+    }
+
+    /// Sort the header directives by weight.
+    ///
+    /// Headers with a higher `q=` value will be returned first. If two
+    /// directives have the same weight, the directive that was declared later
+    /// will be returned first.
+    pub fn sort(&mut self) {
+        sort_by_weight(&mut self.entries);
+    }
+
+    /// Determine the most suitable `Content-Type` encoding.
+    ///
+    /// # Errors
+    ///
+    /// If no suitable encoding is found, an error with the status of `406` will be returned.
+    pub fn negotiate(&mut self, available: &[Encoding]) -> crate::Result<ContentEncoding> {
```

Comment:
It would be much usable if this instead returned `Result<Encoding>`. `tide-compress` is blocked from using this dur to that.

Reply:
Ah it turns out you can `PartialEq` between them. Hmm, probably fine.